### PR TITLE
Updated Raw HID docs to clarify packet/report length

### PR DESCRIPTION
**Pull request by kprinssu**
## Description

The issue stems from the documentation mentioning the following:

```
`raw_hid_receive` can receive variable size packets from host with maximum length `RAW_EPSIZE`.
```

However, the current implementation of the `udi_hid_raw_report_rcvd` is that the incoming report needs to be `UDI_HID_RAW_REPORT_SIZE` otherwise it is ignored and thrown away. @fauxpark pointed out this out as I was having difficulties having my keyboard respond to the data I was sending it.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

N/A

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c [hunk 1]
```diff
@@ -738,7 +738,7 @@ bool udi_hid_raw_receive_report(void) {
 static void udi_hid_raw_report_rcvd(udd_ep_status_t status, iram_size_t nb_rcvd, udd_ep_id_t ep) {
     UNUSED(ep);
 
-    if (status == UDD_EP_TRANSFER_OK && nb_rcvd == UDI_HID_RAW_REPORT_SIZE) {
+    if (status == UDD_EP_TRANSFER_OK && nb_rcvd <= UDI_HID_RAW_REPORT_SIZE) {
         UDI_HID_RAW_RECEIVE(udi_hid_raw_report_recv, UDI_HID_RAW_REPORT_SIZE);
```
Comment (fauxpark) [inline tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c:742]:
```suggestion
    if (status == UDD_EP_TRANSFER_OK) {
        UDI_HID_RAW_RECEIVE(udi_hid_raw_report_recv, (uint8_t)nb_rcvd);
```
This should theoretically make the length parameter you get in `raw_hid_receive()` correct, as it is not in LUFA (it turns out it is always `RAW_EPSIZE` bytes long, and so is the buffer - I'll look into that).

You might also need to do a memset somewhere, to clear out the buffer from previous packets.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c (hunk 1 above)
Comment (kprinssu) [reply tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c]: @fauxpark I am not sure where the best place zero out the buffer as I have tracked down `udi_hid_raw_report_set` to some externally defined function (binary blob?).

I have also noticed that the raw_hid_send transmits junk data and have zero'd the buffer there.

Comment (fauxpark) [reply tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c]: Actually, considering we now have a working length parameter (I hope), maybe we don't really need to clear the buffer...

Comment (kprinssu) [reply tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c]: Marking the conservation resolved then. Please feel free to merge to comment on any other changes,

**Review by drashna — APPROVED** [review]

**Comment by fauxpark** [comment]
I've done some more testing (albeit on the LUFA code as I don't have a Massdrop board), and it seems like HID reports cannot actually be variable length at all - Wireshark tells me my request packet is malformed, unless I send the full 32 (actually 33*) byte HID report. This makes sense since report size is dictated by the descriptor, and AFAIK there is not really any inbuilt way to tell how long a report is supposed to be other than by implementing some kind of protocol that transmits the intended length along with the data. That seems a little out of scope for QMK's raw HID feature.

*See http://hidapi-d.dpldocs.info/hidapi.bindings.hid_write.html

**Comment by yiancar** [comment]
What @fauxpark  said is absolutely true. You MUST send 32bytes as this is what is described by the descriptor.
Some OSes are more relaxed on what happens if you send more or less.
OSX of example rejects anything that is less or more than 32 BUT accepts multiple of 32 and interprets it as separate packets.

TL;DR +1 on fauxpark's comment

**Comment by kprinssu** [comment]
@fauxpark I added an extra check and updated the documentation so it's a bit more verbose.

File: docs/feature_rawhid.md [hunk 2]
```diff
@@ -29,7 +29,7 @@ void raw_hid_receive(uint8_t *data, uint8_t length) {
 }
 ```
 
-`raw_hid_receive` can receive variable size packets from host with maximum length `RAW_EPSIZE`. `raw_hid_send` on the other hand can send packets to host of exactly `RAW_EPSIZE` length, therefore it should be used with data of length `RAW_EPSIZE`.
+`raw_hid_receive` can receive variable size packets from host with maximum length `RAW_EPSIZE` and minimum length of 32 bytes. `raw_hid_send` on the other hand can send packets to host of exactly `RAW_EPSIZE` length, therefore it should be used with data of length `RAW_EPSIZE`.
```
Comment (fauxpark) [inline docs/feature_rawhid.md:32]:
```suggestion
These two functions can receive and send respectively packets from and to the host with length `RAW_EPSIZE` bytes (32 on LUFA/ChibiOS/V-USB, 64 on ATSAM).
```
This entire page kinda needs a bit of rewriting, tbh.

File: tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c [hunk 3]
```diff
@@ -738,8 +738,8 @@ bool udi_hid_raw_receive_report(void) {
 static void udi_hid_raw_report_rcvd(udd_ep_status_t status, iram_size_t nb_rcvd, udd_ep_id_t ep) {
     UNUSED(ep);
 
-    if (status == UDD_EP_TRANSFER_OK && nb_rcvd == UDI_HID_RAW_REPORT_SIZE) {
-        UDI_HID_RAW_RECEIVE(udi_hid_raw_report_recv, UDI_HID_RAW_REPORT_SIZE);
+    if (status == UDD_EP_TRANSFER_OK && nb_rcvd >= 32) {
+        UDI_HID_RAW_RECEIVE(udi_hid_raw_report_recv, (uint8_t)nb_rcvd);
```
Comment (fauxpark) [inline tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c:742]:
```suggestion
    if (status == UDD_EP_TRANSFER_OK && nb_rcvd == UDI_HID_RAW_REPORT_SIZE) {
        UDI_HID_RAW_RECEIVE(udi_hid_raw_report_recv, UDI_HID_RAW_REPORT_SIZE);
```
This basically should be completely reverted. The original behaviour is (unfortunately) correct.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: docs/feature_rawhid.md [hunk 4]
```diff
@@ -29,7 +29,7 @@ void raw_hid_receive(uint8_t *data, uint8_t length) {
 }
 ```
 
-`raw_hid_receive` can receive variable size packets from host with maximum length `RAW_EPSIZE`. `raw_hid_send` on the other hand can send packets to host of exactly `RAW_EPSIZE` length, therefore it should be used with data of length `RAW_EPSIZE`.
+These two functions can receive and send respectively packets from and to the host with length `RAW_EPSIZE` bytes (32 on LUFA/ChibiOS/V-USB, 64 on ATSAM).
```
Comment (fauxpark) [inline docs/feature_rawhid.md:32]:
```suggestion
These two functions send and receive packets of length `RAW_EPSIZE` bytes to and from the host (32 on LUFA/ChibiOS/V-USB, 64 on ATSAM).
```
Sorry, I realised this sentence was a little awkward.

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by fauxpark** [comment]
Thanks!
